### ISSUE TITLE
Fix and improve cleanup script for Round 2

### ISSUE THREAD

**Issue by tschaffter** [Sage-Bionetworks/DigitalMammographyChallenge#387]
- clean docker data thinpool
- clean kernel crash files (save them somewhere else)

Note: the root filesystem of the machines don't have the same size

**Comment by tschaffter** [comment]
Before pulling a docker image:

1. The agent fetches the size of the submitted docker repository (A; repository can be hosted on Synapse but somewhere else). If A > 10 GB, send a message to alert the participant that the docker container is too large.
2. get the free space available to the docker thinpool (B). This can be achieve by running a Bash script that I'll provide or maybe directly from the Java Docker API.
3. If B < 25 GB (2x 1-byte container can grow up to max 10 GB + security margin of 5 GB), run Bash cleanup script to remove image that are not used by running containers. The script can take a few minutes to complete. Not sure if this would be a problem but it would be safer to ensure that the script is not executed (e.g. by the second agent) if it is already running. The Bash cleanup script could create a "lock file" that the second agent could see, which would tell it that he has to wait to do anything as long as the lock file exists. Once the lock file has been removed, the agent on hold can resume with Item 1.
4. Once the first agent has finished to run the cleanup script, he can safely pull the container.

@brucehoff What do you think?

**Comment by brucehoff** [comment]
Some thoughts:  What happens if the agent stops while the script is running?

**Comment by tschaffter** [comment]
@brucehoff Can we use this REST request to fetch information about a repository size?

> GET /containers/(name)/json now accepts a size parameter. Setting this parameter to ‘1’ returns container size information in the SizeRw and SizeRootFs fields.

Source: https://docs.docker.com/engine/reference/api/docker_remote_api/

**Comment by brucehoff** [comment]
@tschaffter 
We can absolutely use the Engine API to check an image's size.  However the goal is to check the size *before* pulling the image [to the Engine].  My hope was to find a *Docker Registry* API that would let us check the size prior to executing the Docker pull command.  Did you have a different strategy in mind?

**Comment by tschaffter** [comment]
Like this one?
https://docs.docker.com/registry/spec/api/

**Comment by brucehoff** [comment]
Yes, that is the documentation for the registry APIs.  But what API is it that tells you the size of an image or layer?   Here is the API for getting an image manifest:
https://docs.docker.com/registry/spec/api/#/pulling-an-image-manifest
(but there is no size information).  Responses to various APIs include a 'Content-Length' header, but it's the size of the JSON metadata, not of the image/layer.

**Comment by tschaffter** [comment]
From what I understand, the layers are stored as gzipped tar files on the registry. Therefore there is no way to know the uncompressed size from the repository. An alternative approach could be to use a separate thread to monitor the amount of data that an agent is trying to pull. The monitoring task should be able to identify the data pulled by the current agent, and not from the second agent if it is pulling at the same time. If the amount of data pulled become larger than 10 GB, then we interrupt the pull task and return an email alert to the participant. What do you think?

**Comment by brucehoff** [comment]
> What do you think?
One way to implement this idea wold be as a modification of the current Docker engine "pull" command to add a size limit.  Another approach would be to let the O/S help out by (somehow) pulling to a file share with (say) a 10GB space limit then moving the successfully downloaded image to the proper location, as expected by Docker.  The Docker "pull" command would break for images that are too large for lack of disk space.

**Comment by tschaffter** [comment]
> One way to implement this idea wold be as a modification of the current Docker engine "pull" command to add a size limit.

This is a way to implement it.

> Another approach would be to let the O/S help out by (somehow) pulling to a file share with (say) a 10GB space limit then moving the successfully downloaded image to the proper location, as expected by Docker.

This is the same approach as for `/modelState`, however I don't think that Docker can store images at different locations on a machine.

**Comment by brucehoff** [comment]
I figured out how to get what seems to be the *compressed* layer size, see https://github.com/Sage-Bionetworks/ChallengeDockerAgent/commit/6f9fed643b2f460f372f0ce0261cab2d67add6da

The caveats are:  I've only tried this on the Synapse Docker registry and also there is a long delay when the code tries to close the input stream of the HTTP Response object (which I only retrieve to check the Content-Length header).  So there HTTP client may be download quite a bit of the layer and I don't know what impact this might have on disk usage.

**Comment by brucehoff** [comment]
Verified compressed size check works with DockerHub as well as Synapse registry: https://github.com/Sage-Bionetworks/ChallengeDockerAgent/commit/ae68b6f20fdc2d59c114bad51276e8c5459a552f

**Comment by tschaffter** [comment]
How do you plan to translate the compressed size of an image into effective size?

**Comment by tschaffter** [comment]
If you send me a list of Docker images that have been submitted during the first round, I can collect their uncompressed size using a script in kb as shown below.

```
[root@bm21 ~]# du -k --max-depth=0 /var/lib/docker/devicemapper/
18575352	/var/lib/docker/devicemapper/
[root@bm21 ~]# docker pull docker.synapse.org/syn4224222/dm-train-caffe
Using default tag: latest
latest: Pulling from syn4224222/dm-train-caffe
8d30e94188e7: Pull complete 
42dc503c149f: Pull complete 
703b89f2b631: Pull complete 
6e90add89e05: Pull complete 
d3f1d75635c9: Pull complete 
360419337e2c: Pull complete 
6534d6abffcf: Pull complete 
9f89baa7ee1d: Pull complete 
887d4d260c5f: Pull complete 
3c39b2ebc485: Pull complete 
97819e52d579: Pull complete 
014191e01c27: Pull complete 
7b9c92158f0d: Pull complete 
28561eb9c751: Pull complete 
02c241d982d9: Pull complete 
5a280ac80a66: Pull complete 
d537d4068ec8: Pull complete 
cd08bf4b5881: Pull complete 
c56e92c91dc2: Pull complete 
0b30db30a870: Pull complete 
850c4bc037ee: Pull complete 
Digest: sha256:c129a9997a3ab440632f773145c6d1edad554b830eb1af9a8e0f0829a4c1ff44
Status: Downloaded newer image for docker.synapse.org/syn4224222/dm-train-caffe:latest
[root@bm21 ~]# du -k --max-depth=0 /var/lib/docker/devicemapper/
22349508	/var/lib/docker/devicemapper/
```

**Comment by brucehoff** [comment]
Using the 'docker images' command I interrogated the Docker Engines on the bare metal machines to get the uncompressed sizes for all the images that are currently downloaded.  I then got the compressed image size information from the registries for these images.  Here are the results:
```
Image	Compressed size	Uncompressed size
library/hello-world@sha256:0256e8a36e2070f7bf2d0b0763dbabdd67798512411de4cdcf9431a1feb60fd9	974	1892.352
docker.synapse.org/syn5644795/dm-directory-clean@sha256:5091763d994cf12db6f3ece7854b6ac282d147cf276449392070bd0729a90b76	125886848	375704780.8
brucehoff/challengedockeragent@sha256:27af7968c774780843116e44be6de6a6c1c3570623c677dd8677fd72fa1ace47	303886675	678428672
docker.synapse.org/syn5644795/nvml-statsd@sha256:5b6f9af77748a8329eed3cae07e528a082e20eabe2d2325918d5244bf3c15084	266464380	707683942.4
docker.synapse.org/syn7895930/dm-cbig-get_info@sha256:6389a72a15ea065fbaf6d3578c4a4d4283b14563ecbb499ed86499196f3d6b1f	335781241	900412211.2
docker.synapse.org/syn7250801/example@sha256:132188f7853789bcf85fe166d900d483175156174d08b6a2491dc1bdaf269627	321173523	938894950.4
docker.synapse.org/syn7250801/example@sha256:54ae0c52bee699d2b6111782bcc4f6c845876e01964d78a785fb05bd504ecd6d	321173513	938894950.4
docker.synapse.org/syn7250801/example@sha256:e38515c054a82530abe71c45bcc0822ef47f522930e90d4178619b46fc60b97e	321173529	938894950.4
docker.synapse.org/syn4224222/dm-preprocess-png@sha256:4a0a19bfff265acfb7a182b6f7866627d7695b2095aeae04f1f52622c2bf57e3	338334654	996147200
docker.synapse.org/syn7247943/preprocess@sha256:3cdc2ae9fccee2ededa9fe5f5d8e262b519ff2b1da342b8a467ec7e0f983ba16	440501112	1192927166
docker.synapse.org/syn7323270/preprocess@sha256:251a101fee7e1b7a96c6540d9d692d9a7a85dbfd167da1e3fa2bc9fbdbdc719b	440526717	1197222134
docker.synapse.org/syn7899960/shreyasnet-0.1.0.5@sha256:db6084e9e36d81fd3d196d1fd7716e1dc745c130d167435369231fae614cc9fb	473186509	1293858898
docker.synapse.org/syn7494640/aneh@sha256:9cd4c77198f5dc891b66aaa1a691264ba31837d6374b1b6d1acfa57d10863d43	408261591	1306743800
docker.synapse.org/syn7860342/dmct_k2030@sha256:c8ad3304ee35c03b0a7aa99bf786136abd146d73db5ef73fbe00c5a8cb595624	518062373	1313186251
docker.synapse.org/syn7250801/example@sha256:152256714ebb9befcfcd38d181ca466b9880df0210200bd484148f6fcf126ea8	539512170	1466731332
docker.synapse.org/syn7250801/example@sha256:2f5f9d859a9c15ef78265dc49f032b22525874966465ee295740c74a1d4ead4a	539512168	1466731332
docker.synapse.org/syn7250801/example@sha256:c3a804899b27b85f63747836a330b2b69be08b0e021b0ea65a4244f2bc1cf1c0	539512153	1466731332
docker.synapse.org/syn7250801/example@sha256:d5996f2eb65b52068b70d81d2ded1aff041af31f8c2ee01dcea2b393ba1ecf58	539512170	1466731332
docker.synapse.org/syn7519427/dnr-preprocessing@sha256:c08e2793f3b23bab87eb7963d0bd103b7f24cc4a5fb35e030ef8b92532df8852	515326962	1647119958
docker.synapse.org/syn7899960/shreyasnet-0.1.0.1@sha256:b9f775c071c787440fced5263283f3e0f7eac463c01e5fde29edfe29d0e3c20c	634767271	1685774664
nvidia/cuda@sha256:a46496a2e78867947e859d96c845e770c7d6d452b79d44463dbc23237e7722dd	920263104	1734093046
docker.synapse.org/syn7500070/dm-scoring@sha256:ac955e4a989d27308f9f9e719ee7525bfe05e5a523aa2901901e7177b919bec3	645776374	2090575331
docker.synapse.org/syn7500070/dm-scoring@sha256:f6faa4dc26011060436c73bb16e4e8aa058711959a93bce691fa1adebf474555	645778957	2090575331
docker.synapse.org/syn7349799/preprocess_base@sha256:f1fb94908888dd57622dfd62fb5702a0758ebba308384b5984aec5e4c6f6a2c8	966007756	2221571834
docker.synapse.org/syn7286737/dm_docker_train@sha256:1e82e67af7ec8eb83a2d7406c25e8ea82902a9860a733fc1ed263708d8c8ba5c	1081925056	2226940543
docker.synapse.org/syn7286737/dm_docker_train_alex@sha256:c7d836a976208fba3d3eecd24b82ff6d11efd04b180dbebd5cef41657cc9e28e	1081925083	2226940543
docker.synapse.org/syn7344325/tmii_2016_12_20_fix_time_ms_500@sha256:aa1f4f40a93b0313e9ed9339bda426fa69a466db258a671b66376164bda5673c	1086968310	2237677961
docker.synapse.org/syn7847533/lenett@sha256:bfc3673583f8998fa95b6c02bfbdbe625ade7c6a60f6c2ec37c9183c9060fdde	1090930745	2247341638
docker.synapse.org/syn7891142/333@sha256:173fd2669799c39f3118863de188aaf092b8d9e28af906d5b2868141b44f49f2	1090930918	2247341638
docker.synapse.org/syn7891142/mursitnett@sha256:8d637f932d14d2d892eb977ef51a358d5eaa57af6cd9a4162b12fd66387e156b	1090930899	2247341638
docker.synapse.org/syn7898481/3-4@sha256:173fd2669799c39f3118863de188aaf092b8d9e28af906d5b2868141b44f49f2	1090930918	2247341638
docker.synapse.org/syn7898481/quarter@sha256:488a9335253d26265c773cf84a254054aa1d2c93e7ddc0303b1b1413ac869784	1090930904	2247341638
docker.synapse.org/syn7221363/dm-training@sha256:2962f9688361e2f693beecb668fed7efa648d9eed8261792c5d41c1eca54263b	1231990089	2512555868
docker.synapse.org/syn7364762/oversampling@sha256:e4af348027fd541bcf7c438e231500bff4231f5812758e2e2a12a389ecafbee1	1197546501	2538325672
docker.synapse.org/syn7364762/smote@sha256:c3eb754d9bb11c3f6d77cdba10461309bd3e74d2ac56bd6a995e8b2bbbb9888d	1197546484	2538325672
docker.synapse.org/syn7519427/dnr-training@sha256:22138b3a61bfe380b2d48a256a39f669071d1a3ed2b7ad5745145758e041b276	1074945141	2583422829
docker.synapse.org/syn7519427/dnr-training@sha256:37a260670c5eaf254ea0af9bec44774cee582e3097841ff81d81d663ebeb8352	1074944953	2583422829
docker.synapse.org/syn7494644/fdg_train@sha256:e263ce026a1aea8e926ab17824ec3fd834b65ead536a7e2328bb9a720d3bd61d	1155292784	2689723269
docker.synapse.org/syn7494640/resnet@sha256:3ee94739053c9451880c1c83b1c74fe6cc2918adb2698fabeedd16e8cede95e4	1158182710	2691870753
docker.synapse.org/syn7494640/resnet@sha256:60f1e902c453aa16c90ddda888344b25b3bdbdf451a31e67ce68f73197199d0a	1158182549	2691870753
docker.synapse.org/syn7494640/training@sha256:4e6cd8e42d287b246dd77b3fccd740ce29e31ca1700b67a83be4e55a14a706d6	1158182955	2691870753
docker.synapse.org/syn7494640/training@sha256:51b696fb5be883a11d9cb7e031113a0020eac83955a92ff08992a8fa44957941	1158183343	2691870753
docker.synapse.org/syn7494640/training@sha256:5b76b453b98174e5fb17527aa58f6c42891ffa897cf3d9148975b572e51c0150	1158183138	2691870753
docker.synapse.org/syn7494640/training@sha256:743c3b5bcedb7a4ecc594ff18bb15f193f8aff1050624c954a87321ec4e50789	1158183024	2691870753
docker.synapse.org/syn7540399/dm_challenge@sha256:1f9d4da6d629751c7d3c1cca640b5cf93ee6c1a92d7057d07f24569cfb1db70a	1593199089	2888365507
docker.synapse.org/syn7540399/dm_challenge@sha256:471a7f269e07837c8805661b406f17ad7c6b9603eee5f99780d0be3bac5d1172	1597836423	2894807958
docker.synapse.org/syn7506112/dmmodel@sha256:86cec57bf396a57b2d1e68b50d581c8b98d4c3bc202a8cfd9d3bbb57b9b42cd4	1337403253	2939905114
docker.synapse.org/syn7506112/dmmodel@sha256:f9483376b3b0b9dbbb5707e0e9960b0c7deb30aea74adbd7eae4f24ef34ed436	1337403458	2939905114
docker.synapse.org/syn7221819/sc1_pre_train_recover@sha256:c4cf24d59cafc52cc9b44f94b570f2cdaa621341de893263f9c7310eb5df8dee	1448054888	3221225472
docker.synapse.org/syn7221819/sc1_pre_train_recover@sha256:ece2f5c610a56604566ed8da23e3e9c1873686035f68fe54f46143055959ab66	1448049658	3221225472
docker.synapse.org/syn7813569/train@sha256:814803978f6dd0feffa956d4971aabecc2020038a709a44decaead9a8f6d79cc	1480791489	3493955895
docker.synapse.org/syn7813569/train@sha256:e77b88eec281d53fdf3d3c4d54aed5b11492159ae84d994c38f65990d5bdcce5	1480791467	3493955895
docker.synapse.org/syn7870507/parallel-preprocessing@sha256:7ae766e21bda1691219acd07faae879fc775b1d3a1cb8031b72d551a14a4f2a7	1613007548	3506840797
docker.synapse.org/syn7870507/parallel-preprocessing@sha256:9455c1614e3dd75da71cf600fe04a4cad0f6ecb4f7999543754ca5e5a4f4643d	1613007547	3506840797
docker.synapse.org/syn7890435/dm-ls-train-dl@sha256:4fdc3fe7f2beab374f133445dcd9b2f0ddd3ae72ed6271ab09631a7625f8d232	1621782817	3530463117
docker.synapse.org/syn7890435/dm-ls-train-dl@sha256:59c56aa95beeafdf5327b39c15c6368816c22278d764f1a451389e416f02d7ce	1621782817	3530463117
docker.synapse.org/syn7890435/dm-ls-train-dl@sha256:b9c00a6dc8ba594e770797759fcdc938c09be84dc4cd009dd825c08c6c89bf92	1621782818	3530463117
docker.synapse.org/syn7799988/mystery_net6@sha256:ba3ae6d517b7028e60d0e74a6508302fd28022096412eaea8a14c26d012994fc	1591215847	3541200536
docker.synapse.org/syn7187934/medsieve@sha256:b4b055f4973dce468f4e247ef3a30a997963f17b7ed32042c2e746e34d515c04	1552902360	3670049554
docker.synapse.org/syn7323270/submit2@sha256:e44081bde89059c7e09a2e7fdf84307e60e7daf35510f392ea186de3a545fae1	2375427910	3782792446
docker.synapse.org/syn4224222/dm-train-caffe@sha256:c129a9997a3ab440632f773145c6d1edad554b830eb1af9a8e0f0829a4c1ff44	1634502944	3812857217
docker.synapse.org/syn4224222/dm-preprocess-caffe@sha256:84e9eac639a43de046ba2c81d102709ba785cf98e414bf75a2847073f2996341	1642528955	3862249341
docker.synapse.org/syn7496088/preprocessing_run4@sha256:07e9a20a0c54f8b4f913fd1671ba3aab0a047414f50409f46bd0e60ceb7b67b5	1403636158	3891240370
docker.synapse.org/syn7716413/sc1_score@sha256:88bab35bd2267cd2f5f3325976c4644c77c56235fb1f80081252a11d5e87d5da	1666977821	3892314112
docker.synapse.org/syn7272726/dm-caffe-pre@sha256:986e464d257de453a73daecd489ac9b7109f37e6d8c340e4999e7695911612b0	1699064268	3976065974
docker.synapse.org/syn7832674/bibo-mxnet-preprocess-crop-png-all@sha256:1364ece5557948c40e7fc481a0759abb3571ac0bcf4580cc146a366c957a293e	1577239022	4025458098
docker.synapse.org/syn7832674/bibo-mxnet-inference-sc1-dream1-50epoch-2@sha256:e26359d1c644fbeb95fd76854d2f811b79c0f66b77306216c39384c433f7b13d	1581592601	4030826807
docker.synapse.org/syn7832674/bibo-mxnet-inference-sc1-dream2-50epoch-2@sha256:8e572e3fe18e77304f2b9a00c5a67a8134eb4e1f7d69f2d35b765c3f317233ea	1581780529	4031900549
docker.synapse.org/syn7887199/basic-training@sha256:5b9d89f2fa42e239e1c25a4b558ebb950bf28d2bbb729d478cd79c8683209b98	1825766494	4083440157
docker.synapse.org/syn7887199/basic-training@sha256:8a0f92abf19a0408eff63d822488ff1d0d0a0e76684d641ccad99c1ac8b27b80	1825767338	4083440157
docker.synapse.org/syn7887199/basic-training@sha256:b6a53ce17ed8e4f3b7ee11962427c9b0fefd90e9bd9bc74f538649dbfb293592	1825766132	4083440157
docker.synapse.org/syn7272726/dm-caffe-infdbg@sha256:1ac82fd316b74917418661a969d6e99a5b6b1bf25fc5af45021211b4d4dcd19c	1916964751	4207994208
docker.synapse.org/syn7272726/dm-caffe-infdbg@sha256:9b09348f68e7967803bea4d3b40be9d173c4cd2e2806ef9dea0f20d59e890e2a	1916964651	4207994208
docker.synapse.org/syn7187934/medsieve@sha256:06574e455f912c1a1f4762547d2effcbcab6d0b6af449ba0efb9ab90917ae094	2086365581	4287451103
docker.synapse.org/syn7187934/medsieve@sha256:0ea1ecdd200dd70a5f56f786f3b582bc72c86a6abb0742ce4ad14abc2db6f23d	2086365584	4287451103
docker.synapse.org/syn7187934/medsieve@sha256:cfb9cd8d642622ecaa3356f533620b79b2923f894ca75761fa9ef9ca90ac3a0a	2086365619	4287451103
docker.synapse.org/syn7187934/medsieve@sha256:e5b9c59722549cb767d935d9684895f18934e97b01d86c623f0ae8425af45286	2086365578	4287451103
docker.synapse.org/syn7295142/4channel@sha256:54ec007589118ea3d37492b13d15b14b1bf70edfc91871cc9ec90ab4c6c7edf5	1808465208	4502199468
docker.synapse.org/syn7295142/4channel@sha256:6d6bf663d318d677b478b8d6f33e541382ebcd86d4dc6df94720b5c98406e98d	1808466452	4502199468
docker.synapse.org/syn7593674/dm-keras-preprocess@sha256:7d927a4f17b04644f12e166a5b55646e50e832586780080919a55dbf98ef9874	2023964549	4727685251
docker.synapse.org/syn7593674/dm-keras-train@sha256:b63a1d61c4c5103337e9c0c7ab1d2648c0ede179b08bc88143063644fcb4b69f	2023964549	4727685251
docker.synapse.org/syn7250801/train@sha256:5193792e773fa4e606e0505db1503320f1c81848e8d1592557f808fef85327db	2040767132	4804994662
docker.synapse.org/syn7342846/training@sha256:3f71b9421b07dfe9845f52b54e55c48ded7f3fa7ab033f798f1c68f4a94a357e	1898083132	4929548714
docker.synapse.org/syn7824993/testbase2@sha256:61f975e218fec44aa2e4f6235dc8f64e7e5187001427ef1c480a5675c658d550	2085431800	5068061409
docker.synapse.org/syn7346568/dm-ls-preprocess@sha256:1cb88dc2dc300e447aa2d922defd780d17298381b008fd744f0056f351b2d1e1	1493155488	6081673691
docker.synapse.org/syn7800554/ivycad_combonet_test_ens15@sha256:38e807ff978cd894b76edd6c0b6d4c491b8ba80bd58abf798753a944e0a55084	7054374338	10748155658
```

**Comment by brucehoff** [comment]
![compression](https://cloud.githubusercontent.com/assets/1627352/21533920/b7907320-cd14-11e6-89ed-18c403fd9ed2.png)

**Comment by tschaffter** [comment]
After updating the kernel and other packages on the SoftLayer machines, here are the steps that I'm taking regarding docker and the data space that it uses.

Below are a list of directories to handle during cleanup:
- /var/lib/docker/tmp
- /var/crash/ (compress to /data/data/crashes/ before removing files from the root FS)
- /root/tempDir

1. Remove all containers and images
```
# docker stop $(docker ps -q)
# docker rm $(docker ps -aq)
# docker rmi $(docker images -aq)
```
2. Clean the content of `/var/lib/docker/tmp` (not counted toward the data space that Docker is keeping track of and returned by `docker info`). In the future, be careful about the content of this directory. Not sure if it can be cleaned when docker is running jobs.
```
# service docker stop
# rm -fr /var/lib/docker/tmp/*
```
3. Clean the content of `/var/crash/` (8 GB at some point on bm08). Move to a different partition (here mounted to /data/data).
```
# mkdir -p /data/data/crashes; mv /var/crash/* /data/data/crashes/.
```
4. Check using `du -h /` that the root FS is hosting less than 10 GB.
5. Run `./DigitalMammographyChallenge/setupNewServer.sh softlayer bmXX` to make sure that the latest version of the docker service is installed (currently 1.12.5, build 7392c3b). This script now exits without install the sysdig and nvml-statsd container (not required for now; should be installed using a separate bash script).
6. Reduce the data space available to Docker to help guaranteeing that the root FS doesn't become full. By default, Docker is allowed to use up to 100 GB of space. The smallest root FS among the 23 machines is 100 GB. Clean root FS is less than 10 GB, however docker, crashes and data generated by the agents generate data that must be regularly cleaned. Let's assume that they will never use more than 10 GB altogether, this leaves 80 GB for Docker.
```
# service docker stop
# rm -fr /var/lib/docker/*
Manually add "--storage-opt dm.loopdatasize=75G" after "--storage-driver devicemapper" to /etc/systemd/system/docker.service. For some reason, Docker allocate 5 GB more than specified.
# systemctl daemon-reload
# service docker start
# # docker info | grep "Data Space Total"
```
7. Comment the `crontab -e` entries related to nvml-statsd and cleanup_docker_data.sh.

NOTE: The folder `/root/tempDir` contains 2.9 GB of data. I believe that the content of this folder is generated by the agents. Make sure that these files are cleanup regularly. This must most likely be included as part of the main cleanup process.

IDEA: prevent agents to run new jobs if disk space issue arise
TODO: test how big docker images > 10 GB can grow
NOTE: didn't clean /preprocessedData, /scratch, etc.

**Comment by brucehoff** [comment]
> The folder /root/tempDir contains 2.9 GB of data. I believe that the content of this folder is generated by the agents.

The -1 and -2 agents on each machine use /root/tempDir0 and /root/tempDir1, respectively.  They *may* have used /root/tempDir in the past, but should not do so going forward.  It should be the agent's responsibility to clean up /root/tempDirX (X=0,1).

**Comment by tschaffter** [comment]
@brucehoff  Can I remove `/root/tempDir`, `/root/tempDir0` and `/root/tempDir1` as part of the maintenance?

**Comment by brucehoff** [comment]
Y s!

**Comment by brucehoff** [comment]
I mean "Yes"!

**Comment by tschaffter** [comment]
- By default, a container can grow to max 10 GB. `docker ps` is hanging when a container has grown to reach its max size, at which point it is hanging for 30 seconds or so before exiting. At that moment `docker ps` returns the result.
- By default, docker can not build an image with more than 10 GB of data. Tried by adding 5G files to it (error occurs when adding second 5G file).

```
Step 6 : COPY preprocess.sh .
 ---> 6b13f2562968
Removing intermediate container 70c90b8183ee
Step 7 : COPY random_5G.img .
 ---> abb654ff7d11
Removing intermediate container aee152f36610
Step 8 : COPY random_5G.img plop
Error processing tar file(exit status 1): write /plop: no space left on device
[root@bm04 dm-preprocess-fill-filesystem]# 
Message from syslogd@bm04 at Dec 30 16:15:20 ...
 kernel:unregister_netdevice: waiting for lo to become free. Usage count = 1
^C
[root@bm04 dm-preprocess-fill-filesystem]# 
```

@brucehoff Can you confirm that
- Docker can successfully pull images larger than 10 GB (using one submitted during Round 1)
- Docker has trouble as soon as it tries to run an image larger than 10GB (what happens?)

**Comment by tschaffter** [comment]
@brucehoff I've found two large zip files during the maintenance cleaning on the FS of bm07 and bm10.
- bm07: /data/3346508.zip (34G)
- bm10: /data/3346183.zip (55G)
The filesystem of these machines is 100 GB and it's important to keep it clean in order to avoid having the FS being filled to 100%. Note that `/data` is still part of the FS where the subdirectories in `/data/.` are mounted on dedicated partitions. I've moved the two zip files to `/data/data/extra`. If they are not used anymore, they should be removed.
```

```

**Comment by brucehoff** [comment]
I'm pretty sure these were left over from my manual migration of preprocessed data from one machine to another.  I have now removed the files.

**Comment by brucehoff** [comment]
> @brucehoff Can you confirm that
> Docker can successfully pull images larger than 10 GB (using one submitted during Round 1)

Yes, I did this on an Amazon EC-2 machine.  Here is the partial output of the `docker images` command, showing that I successfully pulled images larger than 10GB.

```
docker images
REPOSITORY                                     TAG                 IMAGE ID            CREATED             SIZE
docker.synapse.org/syn7890435/dm-ls-train-dl   <none>              a8328f8cfac7        12 days ago         12.29 GB
docker.synapse.org/syn7415408/cpi16            <none>              0d7d302a5b84        12 days ago         17.01 GB
docker.synapse.org/syn7890435/dm-ls-train-dl   <none>              dd684485afa1        13 days ago         12.29 GB
docker.synapse.org/syn7890435/dm-ls-train-dl   <none>              5fbfd2508179        13 days ago         12.29 GB
docker.synapse.org/syn7415408/cpi14            <none>              d225295a2a1c        13 days ago         17.01 GB

```

> Docker has trouble as soon as it tries to run an image larger than 10GB (what happens?)

Nope, it works:

```
 docker run --rm -it docker.synapse.org/syn7415408/cpi16@sha256:3da7dc44d9152cecf31af3755f3424f9fb826f73e6b54de718ff081ffab285a0 /bin/bash
[root@d4e2b4743f37 /]# echo hello world
hello world
[root@d4e2b4743f37 /]# exit
exit

```

**Comment by brucehoff** [comment]
@tschaffter and I discusses this issue and decided that it might be best to focus not on rejecting large images (which are generally prevented by the default Docker installation configuration on the participant's machine) but to actively clean up unused images before pulling a new one.  @tschaffter volunteered to provide a script which I will integrate with the challenge agent.  A key consideration is that the two agents sharing a Docker engine do not collide (i.e. one agent cleans up an image which another has just pulled and plans to run).  There are a number of solutions, the simplest being not to clean up images which were recently pulled. @tschaffter offered to see if the Docker Engine API provides the time stamp for when the image was pulled.

**Comment by tschaffter** [comment]
There are 3 cleanup scripts installed on each machines (crontab entrees).
```
*/20 * * * * /bin/bash /root/dm-scripts/check_base_fs.sh > /dev/null 2>&1
0 * * * * /bin/bash /root/dm-scripts/cleanup_docker_data.sh > /dev/null 2>&1
0 * * * * /usr/bin/mkdir -p /data/data/crash && /usr/bin/mv /var/crash/* /data/data/crash/. > /dev/null 2>&1
```
- `check_base_fs.sh` sends an email alert if the filesystem is going to overlap the space reserved to docker (80 GB). The alert is send 5G before effective overlap. Filesystem grows due to crash files, /root/tempDir* (from agents), docker temps files (not counted toward docker data usage).
- `cleanup_docker_data.sh` cleanup unused docker images when docker uses 70% or more of its reserved data. The crontab interval is reduced to 20 minutes (instead of every hour as shown above) for the Express Lane machines (images are pulled much more often).
- The third crontab entry moves the crash files from the file system to another partition

**Comment by tschaffter** [comment]
@brucehoff Out of curiosity, the missing piece is to know if a 10G+ container can grow. Can you run the following command?

```
 docker run --rm -it docker.synapse.org/syn7415408/cpi16@sha256:3da7dc44d9152cecf31af3755f3424f9fb826f73e6b54de718ff081ffab285a0 /usr/bin/head -c 10G </dev/urandom >myfile
```

**Comment by brucehoff** [comment]
```
docker run --rm -it docker.synapse.org/syn7415408/cpi16@sha256:3da7dc44d9152cecf31af3755f3424f9fb826f73e6b54de718ff081ffab285a0 /usr/bin/head -c 10G </dev/urandom >myfile
the input device is not a TTY
```

**Comment by tschaffter** [comment]
Not sure if it's the problem but can you /bin/bash into the container, and then do `/usr/bin/head -c 10G </dev/urandom >myfile`?

**Comment by brucehoff** [comment]
Yes, that seems to work fine:

```
docker run --rm -it docker.synapse.org/syn7415408/cpi16@sha256:3da7dc44d9152cecf31af3755f3424f9fb826f73e6b54de718ff081ffab285a0 /bin/bash
[root@2a610725912e /]# /usr/bin/head -c 10G </dev/urandom >myfile

[root@2a610725912e /]# 
[root@2a610725912e /]# 
```

**Comment by tschaffter** [comment]
Ok, so a vanilla docker config: a container smaller than 10G can only grow up to 10G. If the initial size of the container is more than 10G, it can grow to become larger than 20G. We don't know if there is a limit to how much it can grow.

Another aspect of the cleanup: do the agents clean regularly the content of /root/tempDir*?

**Comment by brucehoff** [comment]
> do the agents clean regularly the content of /root/tempDir*?

Each agent cleans up /root/tempDirX/<UUID> after using that sub-directory.

**Comment by tschaffter** [comment]
Finally, bm06 and bm16 have been updated by the DST. I've followed the protocol referenced above to cleanup these two predictions machines. When running `./DigitalMammographyChallenge/setupNewServer.sh softlayer bm06`, bm06 got a version of docker more recent (1.12.6) than the other machines (1.12.5). The sysdig and nvml-statsd containers were also stated using `./DigitalMammographyChallenge/setupSysdigAgent.sh softlayer bm0X`.

@brucehoff The maintenance of the SoftLayer machines is now completed. bm06 and bm16 can now be reconnected.
